### PR TITLE
More flexibility in mixing benchmarks and process executions

### LINKED ISSUE #89: Enable user place specific process executions of a number of benchmarks on a single axis
Sometimes it is useful to compare several individual executions of different benchmarks. E.g. 
- Fannkuch Redux, Hotspot, Process execution 1
- Fannkuch Redux, Hotspot, Process execution 2
- Fasta, PyPy, Process execution 1
- Fasta, PyPy, Process execution 1

This does not make sense unless all four run sequences are charted with the same y-limits.  This should work even if the selected process executions come from different Krun results files.

/cc @vext01

### PR DESCRIPTION
**Pull request by snim2**
Allow user to plot specific executions from different benchmarks and Krun files.

The -b CLI switch can now be used repeatedly, in the format:

```
-b machine:benchmark:vm:variant:p_exec
```

e.g.

```
-b bencher3:fannkuch_redux:Hotspot:default-java:1 -b bencher5:fasta:PyPy:default-python:0
```

The `get_data_dictionaries()` function now collects ONLY the data that the user has requested (including outliers, if necessary). It also constructs all subplot titles. `get_data_dictionaries()` skips any executions or benchmarks which have crashed or been skipped. It checks that the data gathered from Krun results files is exactly the data requested on the command line.

The `main()` function now reassembles data from `get_data_dictionaries()` into "pages" which are to be plotted. The user can request that all subplots are placed on a single page with the `--one-page` switch. Otherwise, each benchmark key for each machine is plotted on a single page.

This PR also renames the benchmarking machines and reduces the size of the legend text.

A nice test case to try (notice filenames) is:

``` bash
#4 individual process execs on one page
python plot_krun_results.py -i -f results/bencher3_results_outliers_w500.json.bz2 -f results/bencher5_results_outliers_w500.json.bz2 -b bencher3:fannkuch_redux:Hotspot:default-java:0 -b bencher3:fannkuch_redux:Hotspot:default-java:1 -b bencher5:fasta:PyPy:default-python:0 -b bencher5:fasta:PyPy:default-python:1 --one-page

#4 individual process execs on separate pages
python plot_krun_results.py -i -f results/bencher3_results_outliers_w500.json.bz2 -f results/bencher5_results_outliers_w500.json.bz2 -b bencher3:fannkuch_redux:Hotspot:default-java:0 -b bencher3:fannkuch_redux:Hotspot:default-java:1 -b bencher5:fasta:PyPy:default-python:0 -b bencher5:fasta:PyPy:default-python:1
```

Fixes #89

### CODE REVIEW

**Comment by vext01** [comment]
Hey Sarah,

I will look at this today.

Can you expand upon this:

> The get_data_dictionaries() function now collects ONLY the data that the user has requested (including outliers, if necessary).

Is there now a mode which removes outliers, or do you mean that outliers may or maynot be indicated on the plots?

Thanks

File: plot_krun_results.py [hunk 1]
```diff
@@ -45,6 +45,14 @@
     'nbody':            'N-Body',
 }
 
+# Machine names that need to be reformatted.
```
Comment (vext01) [inline plot_krun_results.py:48]: Is there something we can do to make it more obvious that this is specific to our machine setup? Should this go in a `plot_conf.json` config file for example? At the very least a big `XXX`?

**Comment by snim2** [comment]
As before, the user can ask for all outliers to be plotted with `--with-outliers`, or they can ask for "unique" and "common" outliers to be annotated separately with `--with-unique-outliers`. If neither of those switches are used then outliers are not annotated on the plots.

File: plot_krun_results.py (hunk 1 above)
Comment (snim2) [inline plot_krun_results.py:48]: This is also true of the `BENCHMARKS` dict that you created. I agree a config file would be better, but I think we should deal with that in a separate PR and make all the `ALL_CAPS` globals configurable.

Comment (vext01) [inline plot_krun_results.py:48]: Sounds sensible. I had forgotten about BENCHMARKS.

File: plot_krun_results.py [hunk 2]
```diff
@@ -378,31 +388,196 @@ def read_krun_results_file(results_file):
...
+    in the available Krun results files. Also, we pass back the title text
+    for each subplot.
     """
-    data_dictionary = dict()
+
+    data_dictionary = { 'data': dict(), # Measurement data to be plotted.
+                        'all_outliers': dict(),
+                        'common_outliers': dict(),
+                        'unique_outliers': dict(),
+                       }
```
Comment (vext01) [inline plot_krun_results.py]: Funny indent?

File: plot_krun_results.py [hunk 3]
```diff
@@ -497,26 +681,30 @@ def fatal_error(msg):
     else:
         from matplotlib.backends.backend_pdf import PdfPages
         print 'Saving results to: %s' % options.outfile
-    print (('Charting with sliding window size: %d, xlimits: %s, '
-            'single_exec: %s, ') %
-            (options.window_size, options.xlimits, options.single_exec))
+    print (('Charting with sliding window size: %d, xlimits: %s') %
+            (options.window_size, options.xlimits))
```
Comment (vext01) [inline plot_krun_results.py]: Another funny indent?

File: plot_krun_results.py (hunk 2 above)
Comment (snim2) [inline plot_krun_results.py]: Yeah, it's a space to the right of the opening brace. Do you normally line the braces up or align the closing brace to the first `d` of `data_dictionary`?

File: plot_krun_results.py [hunk 4]
```diff
@@ -80,65 +88,88 @@
...
+                    subplot_titles.append(plot_titles[machine][key][index])
+                    if with_outliers:
+                        page_outlier.append(data_dcts['outliers'][machine][key][index])
+                    else:
+                        page_outlier.append(None)
+                    if unique_outliers:
+                        page_common.append(data_dcts['common_outliers'][machine][key][index])
+                        page_unique.append(data_dcts['unique_outliers'][machine][key][index])
+                    else:
+                        page_common.append(None)
```
Comment (vext01) [inline plot_krun_results.py:139]: Not saying this is wrong, but at a glance it's not clear why we add `Nones` into the lists. Maybe scatter a few more comments? Thoughts?

File: plot_krun_results.py (hunk 2 above)
Comment (vext01) [inline plot_krun_results.py]: I usually have no space after the open brace, then the close brace in line with the open brace. This keeps flake8 happy at-least. Up to you.

**Comment by vext01** [comment]
OK, commented. Nothing obviously wrong.

Will now test by running.

File: plot_krun_results.py (hunk 4 above)
Comment (snim2) [inline plot_krun_results.py:139]: The reason is that the `draw_subplot` and `plot_page` functions are expecting to receive either a list (or lol) or outliers, or `None` (or list of `None`s).

So, yes, this probably needs some comments!

**Comment by vext01** [comment]
Am I doing this right?:

```
vext01@bencher2:~/research/warmup_experiment$ python2.7 mark_outliers_in_json.py -f ~/research/partial_data/wave1/warmup_results-bencher3.json.bz2                                                                                                            
Marking outliers with sliding window size: 200
Loading: /home/vext01/research/partial_data/wave1/warmup_results-bencher3.json.bz2
Writing out: /home/vext01/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2


vext01@bencher2:~/research/warmup_experiment$ python2.7 plot_krun_results.py -f ~/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2 --with-outliers
Saving results to: krun_results.pdf
Charting with sliding window size: 200, xlimits: None
Using matplotlib backend: agg
Loading: /home/vext01/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2
WARNING: Skipping: spectralnorm:JRubyTruffle:default-ruby from bencher3 (no executions)
WARNING: Skipping: richards:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: spectralnorm:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: binarytrees:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: fasta:JRubyTruffle:default-ruby from bencher3 (no executions)
WARNING: Skipping: nbody:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: fasta:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: fannkuch_redux:CPython:default-python from bencher3 (no executions)
WARNING: Skipping: richards:HHVM:default-php from bencher3 (no executions)
Traceback (most recent call last):
  File "plot_krun_results.py", line 710, in <module>
    one_page=options.one_page)
  File "plot_krun_results.py", line 142, in main
    all_outliers.append(data_dcts['outliers'][machine][key])
KeyError: 'outliers'
```

**Comment by snim2** [comment]
Pushed a fix. You may also want to add an `-o my_pdf.pdf` or some such to that invocation.

**Comment by vext01** [comment]
Not sure if this is a regression, but I don't see the 5*sigma when passing `--sigma`.

```
python2.7 plot_krun_results.py -f ~/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2 --with-outliers --sigma
```

nosigma

**Comment by snim2** [comment]
It isn't a regression, it happened because you passed in `--sigma` but you didn't pass in `--mean`. I guess we didn't expect that users would want to plot the sigma on its own. However, there is no real reason why this shouldn't be possible, so I will push a fix.

**Comment by vext01** [comment]
I see. Pulling are trying again.

**Comment by vext01** [comment]
That works nicely.

Now a query about outliers.

```
python2.7 plot_krun_results.py -f ~/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2 --with-unique-outliers
```

outliers

Maybe a bug, maybe a problem with my comprehension: all of these graphs have a unique outlier percentage which is > 0. But I don't see any blue dots on the graph. I also notice that (fictional pseudocode) `non_unique_percent == unique_percent`. Perhaps the non-unique spots are on top of the unique spots? But shouldn't the two me mutually exclusive?

Separate tweaks for another PR:
- Can we make one of the outlier types a cross so they are easier to distinguish from the other outlier type?
- The above graph highlights the "legend is too big" thing that we have seen before. Could make it flow vertical and have a smaller font size?

**Comment by snim2** [comment]
The issue with common and unique outliers was a bug. I have pushed a fix for that, and changed one of the marker styles.

I already reduced the size of the legend text in an earlier commit in this PR. Making the legend "flow vertical" will mean that it covers the data in all cases. Already the legend covers the data in many cases. Fixing this will not be simple (probably the best fix will be to place the legend outside the plot), so I think you need to raise a separate issue for that one.

**Comment by vext01** [comment]
Had a fiddle around with a few different legend placements. How do you fin this?

legend_under

**Comment by snim2** [comment]
That's much improved, I think. I doubt we'll do better.

**Comment by vext01** [comment]
Couple of other questions:
- What is the white line in the above graph? Is it the mean in the wrong colour?
- Should --outliers and --unique-outliers be mutually exclusive?

**Comment by snim2** [comment]
BTW the new legends look a bit odd in interactive mode.

**Comment by vext01** [comment]
Will look into interactive mode.

Another key error:

```
$ python2.7 plot_krun_results.py -f ~/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2 --with-unique-outliers --sigma --mean -b bencher3:binarytrees:C:default-c:0 -b bencher3:fasta:C:default-c:1 --with-outliers --one-page
Saving results to: krun_results.pdf
Charting with sliding window size: 200, xlimits: None
Using matplotlib backend: agg
Loading: /home/vext01/research/partial_data/wave1/warmup_results-bencher3_outliers_w200.json.bz2
Traceback (most recent call last):
  File "plot_krun_results.py", line 728, in <module>
    one_page=options.one_page)
  File "plot_krun_results.py", line 134, in main
    page_outlier.append(data_dcts['outliers'][machine][key][index])
KeyError: 'outliers'
```

**Comment by vext01** [comment]
Ug. When you have two plots on the same page, the legends overlap...

**Comment by vext01** [comment]
FWIW this is the best I could come up with. Pushed commit.

comp

**Comment by snim2** [comment]
The last commit works in both interactive and non-interactive modes.

Interactive mode:

legend_interactive_mode

Saved to disk:

legend_non_interactive

**Comment by vext01** [comment]
Looks good, just two minor quibbles:
- Large gap between top of graph and legend in PDF version
- Missing legend entry for 5 sigma

**Comment by snim2** [comment]
Latest commit (non-interactive):

legend_non_interactive_with_sigma

**Comment by vext01** [comment]
Looks good! Maybe a teensy more space between graphs and legends, but otherwise let's merge!

Squash?

**Comment by snim2** [comment]
OK, will squash next

**Comment by snim2** [comment]
Splat.

**Comment by vext01** [comment]
Yeehaw!

**Pull request merged**